crawl.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	maddrs := make([]multiaddr.Multiaddr, len(peerList))
+
+	for i, bp := range peerList {
+		var err error
+		maddrs[i], err = multiaddr.NewMultiaddr(bp)
+		if err != nil {
+			return nil
+		}
+	}
+
+	peers, _ := peer.AddrInfosFromP2pAddrs(maddrs...)
+
+	return peers
+}
+
+func SetupFilecoinPeer(
+	ctx context.Context,
+	network NetworkName,
+) (*FilecoinPeer, error) {
+	if network == "" {
+		network = Mainnet
+	}
+
+	var pp protocol.ID
+	var bootstrapPeers []peer.AddrInfo
+
+	switch network {
+	case Mainnet:
+		pp = "/fil/kad/testnetnet"
+		bootstrapPeers = FilecoinBootstrapPeers(MainnetPeers)
+	case Calibnet:
+		pp = "/fil/kad/calibnet"
+		bootstrapPeers = FilecoinBootstrapPeers(CalibnetPeers)
+	}
+
+	var ddht *dht.IpfsDHT
+	priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
+
+	if err != nil {
+		return &FilecoinPeer{}, err
+	}
+
+	libp2pOpts := []libp2p.Option{
+		libp2p.ListenAddrStrings(
+			"/ip4/0.0.0.0/tcp/9000",      // regular tcp connections
+			"/ip4/0.0.0.0/udp/9000/quic", // a UDP endpoint for the QUIC transport
+		),
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Security(noise.ID, noise.New),
+		libp2p.DefaultTransports,
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			ddht, err := dht.New(
+				ctx,
+				h,
+				dht.Option(dht.Datastore(dsync.MutexWrap(ds.NewMapDatastore()))),
+				dht.Option(dht.NamespacedValidator("pk", record.PublicKeyValidator{})),
+				dht.Option(dht.NamespacedValidator("ipns", ipns.Validator{KeyBook: h.Peerstore()})),
+				dht.Option(dht.Concurrency(10)),
+				dht.Option(dht.Mode(dht.ModeAuto)),
+				dht.Option(dht.ProtocolPrefix(pp)),
+			)
+
+			return ddht, err
+		}),
+		libp2p.NATPortMap(),
+		libp2p.ConnectionManager(connmgr.NewConnManager(100, 600, time.Minute)),
+		libp2p.EnableAutoRelay(),
+		libp2p.EnableNATService(),
+	}
+
+	h, err := libp2p.New(libp2pOpts...)
+
+	if err != nil {
+		return &FilecoinPeer{}, err
+	}
+
+	return &FilecoinPeer{ctx: ctx, host: h, dht: ddht, bootstrapPeers: bootstrapPeers}, nil
+}
+
+func (f *FilecoinPeer) Bootstrap() {
+	connected := make(chan struct{})
+
+	var wg sync.WaitGroup
+	for _, pinfo := range f.bootstrapPeers {
+		wg.Add(1)
+		go func(p peer.AddrInfo) {
+			defer wg.Done()
+			err := f.host.Connect(f.ctx, p)
+			if err != nil {
+				log.Println(err)
+				return
+			}
+			log.Println("Connected to ", p.ID)
+			connected <- struct{}{}
+		}(pinfo)
+	}
+
+	go func() {
+		wg.Wait()
+		close(connected)
+	}()
+}
+
+func (f *FilecoinPeer) Crawl(d time.Duration) {
+	// connect to bootstrap peers
+	go f.Bootstrap()
+
+	// use ticker instead of sleep to make up for slow receivers
+	stop := make(chan struct{}, 1)
+	stopTicker := time.NewTicker(d)
+
+	statusTicker := time.NewTicker(time.Duration(5) * time.Second)
+
+	for {
+		select {
+		case <-statusTicker.C:
+			log.Printf("found peers: %v\n", len(f.host.Peerstore().Peers()))
+		case <-stopTicker.C:
+			stop <- struct{}{}
+			return
```

Comment:
As a show for our appreciation of the work done, always stop tickers and timers (+drain) when not using them anymore. Otherwise resources are not freed and might be a cause for memory leaks (not here, but it's just a thing to remember to do always).

Reply:
Ah, good to know -- thank you 🙏🏽